### PR TITLE
Thinning from below

### LINKED ISSUE #32: Import thinning model as an operation into sim-workbench
From additional-content repository sim12/sim12.c, port the Thin function as a model function and wrap it as an operation into sim-workbench/forestry. This operation should collect statistics about the operation in the aggregation data structure. See report_volume operation as an example of aggregated data. The statistics could be:
* volume of felled treed
* volume of remaining trees (use the debug volume calculation function since we have no proper one yet)

### PR DESCRIPTION
**Pull request by dmela**
Implemented operation `thinning_from_below` with thinning limits to be solved from a lookup table located in the `thinning_limits` module.

Also needed to modify the collecting of variables from simulator level to the doman operation level. For this created module `aggregate_util` located in `forestry` directory.

closes #32 

made separate tasks for other thinning algorithms located in sim12.c (see. #52 #51 #50)

### CODE REVIEW

File: forestry/forestry_utils.py [hunk 1]
```diff
@@ -11,13 +12,46 @@ def compounded_growth_factor(growth_percent: float, years: int) -> float:
...
+                break
+            # weighted sum
+            dw_sum += d * w
+            n += w
+        # average of weighted sums
+        return dw_sum / n if n > 0 else 0
+
+
+def overall_basal_area(stand: ForestStand) -> float:
+    """ Overall basal area of stand in square centimeters """
```
Comment (taplem) [inline forestry/forestry_utils.py:35]: The code is fine, but should the docstring say "square meters"? Looks like `calculate_basal_area` converts cm to m.

File: forestry/thinning_limits.py [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+        for hdom_i in hdoms:
+            if hdom_x < hdom_i:
+                return hdom_i
+    else:
+        if hdom_x < hdom_0:
+            # under
+            return hdom_0
+        else:
+            # over
+            return hdom_n
```
Comment (taplem) [inline forestry/thinning_limits.py:495]: There's a [`bisect`](https://docs.python.org/3/library/bisect.html#bisect.bisect) function in the standard library that can do this for you:
```python
idx = bisect(hdoms, hdom_x)
return hdoms[min(idx,len(hdoms)-1)]
```

File: forestry/forestry_utils.py (hunk 1 above)
Comment (dmela) [reply forestry/forestry_utils.py]: You are right. Fixed!

File: tests/forestry/forestry_utils_test.py [hunk 3]
```diff
@@ -1,9 +1,56 @@
...
+        self.assertEqual(2.8702, round(result, 4))
+
+    def test_solve_dominant_species(self):
+        stand = ForestStand()
+        diameters = [10.0, 11.0, 12.0]
+        stems = [99, 100, 101]
+        species = [1, 2, 2]
+        stand.reference_trees = [
+            ReferenceTree(breast_height_diameter=d, stems_per_ha=f, species=s)
+            for d, f, s in zip(diameters, stems, species)
```
Comment (jpy-luke) [inline tests/forestry/forestry_utils_test.py:49]: Nice touch with packing the list

File: forestry/thinning_limits.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+                    22: (16.1, 22.7),
+                    24: (16.1, 22.7),
+                    26: (16.1, 22.7)
+                },
+                SpeciesKey.SPRUCE:  {
+                    10: (12.0, 18.0),
+                    12: (12.0, 18.0),
+                    14: (14.1, 18.9),
+                    16: (15.5, 21.9),
+                    18: (16.1, 22.8), #TODO: bug? 22.8->22.7
```
Comment (jpy-luke) [inline forestry/thinning_limits.py:198]: This commentary note will become lost and forgotten here forever. Please make an issue about it and relay to domain experts for assessment.

**Review by jpy-luke — APPROVED** [review]
Approved. Still, see comments.

File: forestry/thinning_limits.py (hunk 4 above)
Comment (dmela) [reply forestry/thinning_limits.py:198]: Good point. Already forgot it myself..

Comment (dmela) [reply forestry/thinning_limits.py:198]: Made a issue.

File: forestry/thinning_limits.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+
+    site_type variable explanations:
+    (1) very rich sites (OMaT in South Finland)
+    (2) rich sites (OMT in South Finland)
+    (3) damp sites (MT in South Finland)
+    (4) sub-dry sites (VT in South Finland)
+    (5) dry sites (CT in South Finland)
+    (6) barren sites (ClT in South Finland)
+    (7) rocky or sandy areas
+    (8) open mountains
```
[the remark this answers is not in the document]
Comment (jpy-luke) [reply forestry/thinning_limits.py:429]: Porting work to use the internal enumerations can't be done before the staging branch is merged and release as a new version. So, later.

[the remark this answers is not in the document]
Comment (dmela) [reply forestry/thinning_limits.py:429]: Ok, let's refactor after these changes have been taken into use in forestdatamodel.

**Pull request merged**